yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,981 @@
+  }
+
+  /**
+   * Get the timeline client.
+   * @return the client
+   * @throws Exception if the timeline client is not currently running
+   */
+  def getTimelineClient: TimelineClient = {
+    timelineClient.getOrElse(throw new Exception("Timeline client not running"))
+  }
+
+  /**
+   * Get the total number of processed events
+   * @return counter of events processed
+   */
+  def getEventsProcessed: Int = {
+    eventsProcessed.get
+  }
+
+  /**
+   * Get the total number of events queued
+   * @return the total event count
+   */
+  def getEventsQueued: Int = {
+    eventsQueued.get
+  }
+
+  /**
+   * Get the current size of the queue
+   * @return the current queue length
+   */
+  def getQueueSize: Int = {
+    actionQueue.size
+  }
+
+  /**
+   * Get the current batch size
+   * @return the batch size
+   */
+  def getBatchSize: Int = {
+    batchSize
+  }
+
+  /**
+   * Get the total number of failed posts events
+   * @return counter of timeline post operations which failed
+   */
+  def getEventPostFailures: Int = {
+    eventPostFailures.get
+  }
+
+  /**
+   * is the asynchronous posting thread active?
+   * @return true if the post thread has started; false if it has not yet/ever started, or
+   *         if it has finished.
+   */
+  def isPostThreadActive: Boolean = {
+    postThreadActive.get
+  }
+
+  /**
+   * The YARN application ID of this history service
+   * @return the application ID provided when the service started
+   */
+  def applicationId: ApplicationId = { _applicationId }
+
+  /**
+   * The YARN attempt ID of this history service
+   * @return the attempt ID provided when the service started
+   */
+  def attemptId: Option[ApplicationAttemptId] = { _attemptId }
+
+  /**
+   * Reset the timeline client
+   * <p>
+   * 1. Stop the timeline client service if running.
+   * 2. set the `timelineClient` field to `None`
+   */
+  def stopTimelineClient(): Unit = {
+    stopOptionalService(timelineClient)
+    timelineClient = None
+  }
+
+  /**
+   * Create the timeline domain
+   * @return a domain string or null
+   */
+  private def createTimelineDomain(): String = {
+    val sparkConf = sparkContext.getConf
+    val aclsFlag = sparkConf.getOption("spark.acls.enable")
+    val aclsOn = aclsFlag.getOrElse(
+      sparkConf.get("spark.ui.acls.enable", "false")).toBoolean
+    if (!aclsOn) {
+      logDebug("ACLs are disabled; not creating the timeline domain")
+      return null
+    }
+    val predefDomain = sparkConf.getOption(TIMELINE_DOMAIN)
+    if (predefDomain.isDefined) {
+      return predefDomain.get
+    }
+    val current = UserGroupInformation.getCurrentUser.getShortUserName
+    val adminAcls  = stringToSet(sparkConf.get("spark.admin.acls", ""))
+    val viewAcls = stringToSet(sparkConf.get("spark.ui.view.acls", ""))
+    val modifyAcls = stringToSet(sparkConf.get("spark.modify.acls", ""))
+
+    val readers = (adminAcls ++ modifyAcls ++ viewAcls).foldLeft(current)(_ + " " + _)
```

Comment:
minor, but this feels a little easier to read to me:

    (Seq(current) ++ adminAcls ++ modifyAcls ++ viewAcls).mkString(" ")